distribution/xfer/download.go

```diff
@@ -319,11 +319,15 @@ func (ldm *LayerDownloadManager) makeDownloadFunc(descriptor DownloadDescriptor,
 				return
 			}
 
-			var src *distribution.Descriptor
-			if fs, ok := descriptor.(layer.ForeignSourcer); ok {
-				src = fs.ForeignSource()
+			var src distribution.Descriptor
+			if fs, ok := descriptor.(distribution.Describable); ok {
+				src = fs.Descriptor()
+			}
+			if ds, ok := d.layerStore.(layer.DescribableStore); ok {
+				d.layer, err = ds.RegisterWithDescriptor(inflatedLayerData, parentLayer, src)
+			} else {
+				d.layer, err = d.layerStore.Register(inflatedLayerData, parentLayer)
 			}
```

Comment:
It might be worth factoring this into its own function. I'd like to avoid code duplication between the two makeDownload* functions as much as is practical.